docs/markdown/Using Pants/environments.md

```diff
@@ -0,0 +1,221 @@
+  ..
+)
+
+local_environment(
+  name="local",
+  compatible_platforms=["linux_x86_64"],
+  fallback_environment="docker",
+)
+
+docker_environment(
+  name="docker",
+  ..
+)
+```
+
+In future versions, environment targets will gain additional predicates to control whether they match (for example: `local_environment` will likely gain a predicate that looks for the [presence or value of an environment variable](https://github.com/pantsbuild/pants/issues/17107). But in the meantime, it's possible to override which environments are matched for particular use cases by overriding their configured names: see the "Toggle use of an environment" workflow below for an example.
+
+## Example workflows
+
+### Enabling remote execution globally
+
+`remote_environment` targets match unless the [`--remote-execution`](doc:reference-global#remote_execution) option is disabled. So to cause a particular environment name to use remote execution whenever it is enabled, you could define environment targets which tried remote execution first, and then fell back to local execution:
+
+```python
+remote_environment(
+  name="remote_busybox",
+  platform="linux_x86_64",
+  extra_platform_properties={"container-image=busybox:latest"},
+  fallback_environment="local",
+)
+
+local_environment(
+  name="local",
+  compatible_platforms=[...],
+)
+```
+
+You'd then give your `remote_environment` target an unassuming name like "default":
+```toml
+[environments.names]
+default = "//:remote_busybox"
+local = "//:local"
+```
+... and use that environment by default with all targets. Users or consumers like CI could then toggle whether remote execution was used by setting `--remote-execution`.
+
+> 🚧 Speculation of remote execution
+> 
+> The `2.15.x` series of Pants does not yet support ["speculating" remote execution](https://github.com/pantsbuild/pants/issues/8353) by racing it against another environment (usually local or docker). While we expect that this will be necessary to make remote execution a viable option for local execution on user's laptops (where network connections are less reliable), it is less critical for CI use-cases.
+
+### Use a `docker_environment` to build the inputs to a `docker_image`
+
+To build a `docker_image` target containing a `pex_binary` which uses native (i.e. compiled) dependencies on a `macOS` machine, you can configure the `pex_binary` to be built in a `docker_environment`.
+
+You'll need a `docker_environment` which uses an image containing the relevant build-time requirements of your PEX. At a minimum, you'll need Python itself:
+```python
+docker_environment(
+  name="python_bullseye",
+  platform="linux_x86_64",
+  image="python:3.9.14-slim-bullseye@sha256-abcd123...",
+  ..
+)
+```
+
+Next, mark your `pex_binary` target with this environment (with the name `python_bullseye`: see "Defining environments" above), and define a `docker_image` target depending on it.
+
+```python
+pex_binary(
+  name="main",
+  environment="python_bullseye",
```

Comment:
Is it possible to `parameterize` an `environment`? For this example specifically, I'm typically calling `./pants package` on the `docker_image` target and the PEX is an implementation detail (therefore a Linux platform is what I want), but sometimes I want to run a PEX directly so I'll `./pants package` it expecting a Mac-compatible output.

Reply:
I guess that is also achievable by overriding the environment name to point at my local env when I'm building the pex directly 🤔

Reply:
It can be, and we should probably document that.

Reply:
Added a section.